### ISSUE TITLE
Raising exception in callback works with Popen but not with check_output

### ISSUE THREAD

**Issue by jikamens** [aklajnert/pytest-subprocess#16]
If I add this test to `tests/test_subprocess.py`:

```
def test_raise_exception_check_output(fake_process):
    def callback_function(process):
        process.returncode = 1
        raise PermissionError("exception raised by subprocess")

    fake_process.register_subprocess(["test"], callback=callback_function)

    with pytest.raises(PermissionError, match="exception raised by subprocess"):
        process = subprocess.check_output(["test"])

    assert process.returncode == 1
```

The test fails.

This should behave the same as the code in `test_raise_exception` does, shouldn't it?

**Comment by aklajnert** [comment]
The internal API of `subprocess` checks if the process has returned non-zero exit code, and raises the `subprocess.CalledProcessError`, regardless of what was the cause of an error.  
I don't think that changing this behavior is a good idea, as it should be as consistent with real `subprocess` as possible.

**Comment by jikamens** [comment]
Let me explain my goal a bit more clearly and why I believe this behavior is wrong.

In real life, external executables invoked by the subprocesses module can't raise exceptions. If you call something in subprocess and get back an exception, that's the subprocess module raising the exception, not the external process that it invoked.

Nevertheless, sometimes in unit tests we need to mock out those exceptions that the subprocess module can raise, to confirm that the code we're testing handles those exceptions properly. I thought that was what the recent changes you implemented to support exceptions was intended to do.

What this bug is about is that that it doesn't seem to work in all cases.

Let me illustrate by way of example.

I have code that invokes an external executable. It has to deal with the case that the executable it's invoking does not exist. If that happens in real life, then it will get back a FileNotFoundError:

```
>>> subprocess.check_output('no-such-executable')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/lib/python3.8/subprocess.py", line 411, in check_output
    return run(*popenargs, stdout=PIPE, timeout=timeout, check=True,
  File "/usr/lib/python3.8/subprocess.py", line 489, in run
    with Popen(*popenargs, **kwargs) as process:
  File "/usr/lib/python3.8/subprocess.py", line 854, in __init__
    self._execute_child(args, executable, preexec_fn, close_fds,
  File "/usr/lib/python3.8/subprocess.py", line 1702, in _execute_child
    raise child_exception_type(errno_num, err_msg, err_filename)
FileNotFoundError: [Errno 2] No such file or directory: 'no-such-executable'
```

I need to be able to emulate that in a test. Right now I can't do that with pytest-subprocess. That's what this bug is about.

**Comment by aklajnert** [comment]
OK, I agree that it makes sense. Since there is possibility to deliberately raise an exception from the callback code, it should be respected - I see few other potential benefits of that, even if that's not entirely in line with real subprocess behavior.  
When I looked into the `subprocess.check_output()` code, I thought that I would need to patch whole `subprocess.run()` to make it work which would greatly complicate the code, but I've found better solution on how to implement this.
